c_src/notif_set_tlh_htable.c

```diff
@@ -0,0 +1,198 @@
+
+struct Notif_set *notif_set_new()
+{
+    struct Notif_set *tlh = malloc(sizeof(struct Notif_set));
+    if(tlh == NULL) {
+        return NULL;
+    }
+    htable_init(&tlh->lev1);
+    tlh->len = 0;
+    return tlh;
+}
+
+int notif_set_destroy(struct Notif_set *tab)
+{
+    struct htable_bucket *itr;
+    // destroy each level2 table
+    for(itr = htable_next(&tab->lev1, NULL); itr; itr = htable_next(&tab->lev1, itr)) {
+        htable_reset((struct htable*) itr->value);
+    }
+    htable_reset(&tab->lev1);
+    free(tab);
+
+    return 1;
+}
+
+int notif_set_insert(struct Notif_set *tab, struct Message *msg)
+{
+    // IMPORTANT: Each htable owns its keys, but the 
+    //            values are owned by the user.
+    if(msg == NULL) {
+        return 0;
+    }
+    struct Key *key = htable_key_int(msg->wd);
+    struct htable_ret get_ret = htable_get(&tab->lev1, key);
+    struct htable *lev2_tab = (struct htable*) get_ret.value;
+    if(!get_ret.ok) {
+        // We have a new directory wd; insert to level1
+        // First create a level2 table for this wd (the value)
+        lev2_tab = malloc(sizeof(struct htable));
+        assert(lev2_tab != NULL);
+        htable_init(lev2_tab);
+        htable_put(&tab->lev1, key, (uint64_t) lev2_tab);
+    }
+    else {
+        htable_key_free(key); //because did not put it in the htable
+    }
+
+    // add the filename to the level2 hash table of this wd
+    key = htable_key_str(msg->name);
+    if(!htable_put(lev2_tab, key, 0).ok) { // TODO: store len as well
+        htable_key_free(key); //because did not put it in the htable
+        return 0;
+    }
+
+    tab->len++;
+    return 1;
+}
+
+int notif_set_find(struct Notif_set *tab, struct Message *msg, void *val)
+{
+    (void) val; // To eliminate unused variable warning
+    if(msg == NULL) {
+        return 0;
+    }
+    struct Key *key = htable_key_int(msg->wd);
+    struct htable_ret get_ret = htable_get(&tab->lev1, key);
+    htable_key_free(key);
+    if(!get_ret.ok) {
+        return 0;
+    }
+    key = htable_key_str(msg->name);
+    get_ret = htable_get((struct htable*) get_ret.value, key);
+    htable_key_free(key);
+    if(!get_ret.ok) {
+        return 0;
+    }
+
+    return 1;
+}
+
+struct Notif_set_itr *notif_set_itr_init(struct Notif_set *tab)
+{
+    struct Notif_set_itr *itr = calloc(1, sizeof(*itr));
+    assert(itr != NULL);
+    if(tab != NULL) {
+        itr->tab = tab;
+        itr->itr1 = htable_next(&tab->lev1, NULL);
+    }
+
+    return itr;
+}
+
+void notif_set_itr_destroy(struct Notif_set_itr *itr)
+{
+    free(itr);
+}
+
+int notif_set_itr_next(struct Notif_set_itr *itr, struct Message *msg)
+{
+    if(itr->itr1 == NULL) {
+        return 0;
+    }
+    // Advance the second-level iterator first
+    if((itr->itr2 = htable_next((struct htable *) itr->itr1->value, itr->itr2))) {
+        assert(itr->itr1->key->n == sizeof(int));
+        msg->wd = *((int*) itr->itr1->key->bytes); //Really don't like this
```

Comment:
I'd recommend creating a function that takes a key and returns the proper type of it. Will give you a chance of doing a bit of validation.

Reply:
I am not sure how this can be done correctly. Do you mean checking the number of bytes and deciding based on that?